src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs

```diff
@@ -15,71 +16,125 @@ internal sealed partial class Checksum : IObjectWritable, IEquatable<Checksum>
     {
         public static readonly Checksum Null = new Checksum(Array.Empty<byte>());
 
-        private readonly byte[] _checkSum;
-        private int _lazyHash;
-
-        public Checksum(byte[] checksum)
+        [StructLayout(LayoutKind.Explicit, Size = 20)]
+        private struct Sha1Hash : IEquatable<Sha1Hash>
         {
-            // 0 means it is not initialized
-            _lazyHash = 0;
-            _checkSum = checksum;
-        }
+            [FieldOffset(0)]
+            public unsafe fixed byte Value[20];
```

Comment:
all this make  total sense, but it's worth having some comments to help explain what you're doing here with the fields for the next reader.